storage-proofs-update/src/poseidon/vanilla.rs

```diff
@@ -1,111 +1,86 @@
 use std::marker::PhantomData;
 
+use ff::PrimeFieldBits;
+use filecoin_hashers::{Hasher, PoseidonArity};
 use serde::{Deserialize, Serialize};
-use storage_proofs_core::{
-    error::Result,
-    merkle::{MerkleProof, MerkleTreeTrait},
-    proof::ProofScheme,
-};
+use storage_proofs_core::{error::Result, merkle::MerkleProof, proof::ProofScheme};
 
 use crate::{
     constants::{TreeRDomain, TreeRHasher},
     PrivateInputs, PublicParams, SetupParams,
 };
 
 #[derive(Clone, Serialize, Deserialize)]
-pub struct PublicInputs {
-    pub comm_r_old: TreeRDomain,
-    pub comm_d_new: TreeRDomain,
-    pub comm_r_new: TreeRDomain,
+pub struct PublicInputs<F> {
+    #[serde(bound(serialize = "TreeRDomain<F>: Serialize"))]
+    #[serde(bound(deserialize = "TreeRDomain<F>: Deserialize<'de>"))]
+    pub comm_r_old: TreeRDomain<F>,
+    #[serde(bound(serialize = "TreeRDomain<F>: Serialize"))]
+    #[serde(bound(deserialize = "TreeRDomain<F>: Deserialize<'de>"))]
+    pub comm_d_new: TreeRDomain<F>,
+    #[serde(bound(serialize = "TreeRDomain<F>: Serialize"))]
+    #[serde(bound(deserialize = "TreeRDomain<F>: Deserialize<'de>"))]
+    pub comm_r_new: TreeRDomain<F>,
     // The number of high bits to take from each challenge's bits. Used to verify replica encoding
     // in the vanilla proof. `h` is only a public-input for the vanilla proof; the circuit takes
     // `h_select` as a public-input rather than `h`.
     pub h: usize,
 }
 
-#[derive(Serialize, Deserialize)]
-pub struct ChallengeProof<TreeR>
-where
-    TreeR: MerkleTreeTrait<Hasher = TreeRHasher>,
-{
-    #[serde(bound(
-        serialize = "MerkleProof<TreeRHasher, TreeR::Arity, TreeR::SubTreeArity, TreeR::TopTreeArity>: Serialize",
-        deserialize = "MerkleProof<TreeRHasher, TreeR::Arity, TreeR::SubTreeArity, TreeR::TopTreeArity>: Deserialize<'de>"
-    ))]
-    pub proof_r_old:
-        MerkleProof<TreeRHasher, TreeR::Arity, TreeR::SubTreeArity, TreeR::TopTreeArity>,
-    #[serde(bound(
-        serialize = "MerkleProof<TreeRHasher, TreeR::Arity, TreeR::SubTreeArity, TreeR::TopTreeArity>: Serialize",
-        deserialize = "MerkleProof<TreeRHasher, TreeR::Arity, TreeR::SubTreeArity, TreeR::TopTreeArity>: Deserialize<'de>"
-    ))]
-    pub proof_d_new:
-        MerkleProof<TreeRHasher, TreeR::Arity, TreeR::SubTreeArity, TreeR::TopTreeArity>,
-    #[serde(bound(
-        serialize = "MerkleProof<TreeRHasher, TreeR::Arity, TreeR::SubTreeArity, TreeR::TopTreeArity>: Serialize",
-        deserialize = "MerkleProof<TreeRHasher, TreeR::Arity, TreeR::SubTreeArity, TreeR::TopTreeArity>: Deserialize<'de>"
-    ))]
-    pub proof_r_new:
-        MerkleProof<TreeRHasher, TreeR::Arity, TreeR::SubTreeArity, TreeR::TopTreeArity>,
-}
-
-// Implement `Clone` by hand because `MerkleTreeTrait` does not implement `Clone`.
```

Comment:
Note: can remove the `Clone` impl because we removed the `TreeR: MerkleTreeTrait` trait bound.